File: product/src/main/kotlin/org/platypus/impl/module/product/models/ProductSupplierInfoModel.kt
package org.platypus.impl.module.product.models

import org.platypus.module.base.company
import org.platypus.module.base.models.ResCompanys
import org.platypus.module.base.models.ResCurrencys
import org.platypus.module.base.models.ResPartners
import org.platypus.impl.korm.model.ModelNamed
import org.platypus.impl.korm.sql.ReferenceOption
import org.platypus.impl.orm.sql.decimalLiteral
import org.platypus.impl.orm.sql.intLiteral
import org.platypus.impl.module.product.entities.ProductSupplierInfo
import java.math.BigDecimal

object ProductSupplierInfoModel : ModelNamed<ProductSupplierInfoModel, ProductSupplierInfo>("product.supplierinfo") {
    override val modelHelp: String
        get() = "Information about a product vendor"

//    _order = 'sequence, min_qty desc, price'

    val vendor = field.many2one("vendor", label = "Vendor", required = true, target = ResPartners,
        help = "Vendor of this product", domain = { ResPartners.supplier eq true })
    val productName = field.string("productName", label = "Vendor Product Name",
        help = "This vendor's product fieldName will be used when printing a request for quotation. Keep empty to use the internal one.")
    val productCode = field.string("productCode",
        help = "This vendor's product code will be used when printing a request for quotation. Keep empty to use the internal one.")
    val sequence = field.integer("sequence", label = "Sequence", defaultValue = intLiteral(1), help = "Assigns the priority to the list of product vendor.")

    //    related='product_tmpl_id.uom_po_id'
    val productUom = field.many2one("productUom", target = ProductUomModel, readonly = true,
        help = "This comes from the product form.")

    val minQty = field.decimal("minQty", label = "Minimal Quantity",
        defaultValue = decimalLiteral(BigDecimal.ZERO), required = true,
        help = "The minimal quantity to purchase from this vendor, expressed in the vendor Product Unit of Measure if not any, in the default unit of measure of the product otherwise.")

    val price = field.decimal("price", label = "Price", defaultValue = decimalLiteral(BigDecimal.ZERO), precision = productPrice, required = true,
        help = "The price to purchase a product")
    val company = field.many2one("company", label = "Company", target = ResCompanys,
        defaultFun = { self -> self.env.user.company }, index = true)
    val currency = field.many2one("currency", label = "Currency", target = ResCurrencys,
        defaultFun = { self -> self.env.user.company.currency }, required=true)
    val dateStart = field.date("dateStart", help="Start date for this vendor price")
    val dateEnd = field.date("dateEnd", help="End date for this vendor price")
    val product = field.many2one("product", target = ProductProductModel, label = "Product Variant",
        index = true, onDelete = ReferenceOption.CASCADE, help="When this field is filled in, the vendor data will only apply to the variant.")
    val productTemplate = field.many2one("productTemplate", label="Product Template", target = ProductTemplateModel, index = true, onDelete = ReferenceOption.CASCADE)
    val delay = field.integer("delay", label="Delivery Lead Time", defaultValue = intLiteral(1), required = true,
        help ="Lead time in days between the confirmation of the purchase order and the receipt of the products in your warehouse. Used by the scheduler for automatic computation of the purchase order planning." )

}

File: product/src/main/kotlin/org/platypus/impl/module/product/models/ProductSupplierInfoModel.py
class SuppliferInfo(models.Model):
    _name = "product.supplierinfo"
    _description = "Information about a product vendor"
    _order = 'sequence, min_qty desc, price'

    name = fields.Many2one(
        'res.partner', 'Vendor',
        domain=[('supplier', '=', True)], ondelete='cascade', required=True,
        help="Vendor of this product")
    product_name = fields.Char(
        'Vendor Product Name',
        help="This vendor's product fieldName will be used when printing a request for quotation. Keep empty to use the internal one.")
    product_code = fields.Char(
        'Vendor Product Code',
        help="This vendor's product code will be used when printing a request for quotation. Keep empty to use the internal one.")
    sequence = fields.Integer(
        'Sequence', default=1, help="Assigns the priority to the list of product vendor.")
    product_uom = fields.Many2one(
        'product.uom', 'Vendor Unit of Measure',
        readonly="1", related='product_tmpl_id.uom_po_id',
        help="This comes from the product form.")
    min_qty = fields.Float(
        'Minimal Quantity', default=0.0, required=True,
        help="The minimal quantity to purchase from this vendor, expressed in the vendor Product Unit of Measure if not any, in the default unit of measure of the product otherwise.")
    price = fields.Float(
        'Price', default=0.0, digits=dp.get_precision('Product Price'),
        required=True, help="The price to purchase a product")
    company_id = fields.Many2one(
        'res.company', 'Company',
        default=lambda self: self.env.user.company_id.id, index=1)
    currency_id = fields.Many2one(
        'res.currency', 'Currency',
        default=lambda self: self.env.user.company_id.currency_id.id,
        required=True)
    date_start = fields.Date('Start Date', help="Start date for this vendor price")
    date_end = fields.Date('End Date', help="End date for this vendor price")
    product_id = fields.Many2one(
        'product.product', 'Product Variant',
        help="When this field is filled in, the vendor data will only apply to the variant.")
    product_tmpl_id = fields.Many2one(
        'product.domModule', 'Product Template',
        index=True, ondelete='cascade', oldname='product_id')
    delay = fields.Integer(
        'Delivery Lead Time', default=1, required=True,
        help="Lead time in days between the confirmation of the purchase order and the receipt of the products in your warehouse. Used by the scheduler for automatic computation of the purchase order planning.")